File: top_interview_questions/2_medium/9_others/evaluate_reverse_polish_notation/solution.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = list()
        result = None

        for token in tokens:
            if token not in ["*", "/", "+", "-"]:
                stack.append(token)
                continue
            
            operator = token
            operand_two = int(stack.pop())
            operand_one = int(stack.pop())
            
            if operator == "+":
                result = operand_one + operand_two
            elif operator == "-":
                result = operand_one - operand_two
            elif operator == "*":
                result = operand_one * operand_two
            elif operator == "/":
                result = operand_one / operand_two

            stack.append(int(result))
        
        if stack:
            return stack.pop()

        return result

File: top_interview_questions/2_medium/9_others/evaluate_reverse_polish_notation/solution.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	fmt.Println(evalRPN([]string{"4", "13", "5", "/", "+"}))
}

type stack []int

func (s stack) Push(v int) stack {
	return append(s, v)
}

func (s stack) Pop() (stack, int) {
	l := len(s)
	return s[:l-1], s[l-1]
}

func evalRPN(tokens []string) int {
	var s stack
	var operator string
	var r, operandOne, operandTwo int

	operators := map[string]bool{
		"+": true,
		"-": true,
		"*": true,
		"/": true,
	}

	for _, token := range tokens {
		if !operators[token] {
			n, _ := strconv.Atoi(token)
			s = s.Push(n)
			continue
		}

		operator = token
		s, operandTwo = s.Pop()
		s, operandOne = s.Pop()

		switch operator {
		case "+":
			r = operandOne + operandTwo
		case "-":
			r = operandOne - operandTwo
		case "*":
			r = operandOne * operandTwo
		case "/":
			r = operandOne / operandTwo
		}
		fmt.Printf("%d%s%d=%d\n", operandOne, operator, operandTwo, r)
		s = s.Push(r)
	}

	if len(s) > 0 {
		s, r = s.Pop()
	}

	return r
}
